sdk/rust/oak/src/http/util.rs

```diff
@@ -0,0 +1,114 @@
+//
+// Copyright 2020 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::{
+    io::{Receiver, ReceiverExt, Sender, SenderExt},
+    proto::oak::invocation::HttpInvocation,
+    OakError,
+};
+use log::error;
+use oak_abi::label::Label;
+use oak_services::proto::oak::encap::{HttpRequest, HttpResponse};
+
+/// A pair of temporary channels to pass the HTTP request to the Oak Node and receive the response.
+pub struct Pipe {
+    http_invocation: HttpInvocation,
+    pub request_writer: Sender<HttpRequest>,
+    pub response_reader: Receiver<HttpResponse>,
+}
+
+impl Pipe {
+    pub fn new(request_label: &Label, response_label: &Label) -> Result<Self, OakError> {
+        // Create a channel for passing HTTP requests to the HTTP Client node.
+        let (request_writer, request_reader) =
+            crate::io::channel_create("HTTP request", request_label)?;
+
+        let (response_writer, response_reader) =
+            crate::io::channel_create("HTTP response", response_label)?;
+
+        let http_invocation = HttpInvocation {
+            receiver: Some(request_reader),
+            sender: Some(response_writer),
+        };
+
+        Ok(Pipe {
+            http_invocation,
+            request_writer,
+            response_reader,
+        })
+    }
+
+    /// Inserts the incoming HTTP request in the `request channel` part of the `HttpInvocation`.
+    pub fn insert_message(&self, request: HttpRequest) -> Result<(), OakError> {
+        // Put the HTTP request message inside the per-invocation request channel.
+        self.request_writer.send_with_downgrade(&request)
```

Comment:
Just checking: are all these downgrades necessary? If so, it would be good to have a comment, downgrading is really a special operation, and we should be very explicit about it now that we have separate ABI calls.

Reply:
If a node wanted to send a message to an HTTP client node while downgrading, it should be very explicit about it. Definitely we should not hardcode downgrading operations "just in case" a node may need it.